docs: Add Worklets page (#6257)
software-mansion/react-native-reanimated: packages/docs-reanimated/docs/guides/worklets.mdx

---
id: worklets
sidebar_label: Worklets
sidebar_position: 1
---

# Worklets

> Worklets are short-running JavaScript functions that can be run on the [UI thread](/docs/fundamentals/glossary#ui-thread). Reanimated uses worklets to update view properties and react to events on the UI thread.

Reviewer: 1. Not only on the UI thread (also on different threads), but mainly in other JS runtimes than the React runtime.
2. Reanimated doesn't use worklets to update view properties, this is handled mostly by native code. Reanimated uses worklets to calculate view styles for the current frame. Also, Reanimated can run worklets as a reaction to some event while still on the UI thread.

> Worklets are short-running JavaScript functions that can be run on the [UI thread](/docs/fundamentals/glossary#ui-thread). Reanimated uses worklets to update view properties and react to events on the UI thread.

Reviewer 2: What does `short-running` mean? I'd just keep `short` here.

> You can create your own worklets using the `"worklet";` directive at the top of a function.

Reviewer: `"worklet";` or `'worklet';`?

Author: depends on your prettier config

Reviewer 2: We use single quotes in our codebase so let's stick with it here as well.

```javascript
function myWorklet() {
  'worklet';

>   console.log('Running on the UI thread');

Reviewer: ```suggestion
  console.log('Hello from a worklet');
```

}
```

> The [Reanimated Babel Plugin](https://github.com/software-mansion/react-native-reanimated/blob/main/packages/react-native-reanimated/plugin/README-dev.md#basics) looks for functions marked with the `"worklet"` directive and coverts them into serializable objects. We call this process [workletization](/docs/fundamentals/glossary#to-workletize). These objects can then be copied and run over on the UI thread.

Reviewer 3: ```suggestion
The [Reanimated Babel Plugin](https://github.com/software-mansion/react-native-reanimated/blob/main/packages/react-native-reanimated/plugin/README-dev.md#basics) looks for functions marked with the `'worklet'` directive and converts them into serializable objects. We call this process [workletization](/docs/fundamentals/glossary#to-workletize). These objects can then be copied and run over on the UI thread.
```

Most of the time when working with Reanimated and [Gesture Handler](https://docs.swmansion.com/react-native-gesture-handler/) the code is automatically workletized and run on the UI thread by default.

```javascript
import { useAnimatedStyle } from 'react-native-reanimated';

function App() {
  const style = useAnimatedStyle(() => {
    // Running on the UI thread
    return { opacity: 0.5 };
  });
}
```

> The `"worklet";` directive has no effect on the [JavaScript thread](/docs/fundamentals/glossary#javascript-thread). Worklets on the JavaScript thread are just functions.

Reviewer: In fact, Reanimated Babel plugin breaks hoisting also on the JS thread.

> The `"worklet";` directive has no effect on the [JavaScript thread](/docs/fundamentals/glossary#javascript-thread). Worklets on the JavaScript thread are just functions.

Reviewer 3: Also
```suggestion
The `'worklet';` directive has no effect on the [JavaScript thread](/docs/fundamentals/glossary#javascript-thread). Worklets on the JavaScript thread are just functions.
```

> You can use [`runOnUI`](/docs/threading/runOnUI) to manually run worklets on the UI thread:

Reviewer 2: ```suggestion
You can use [`runOnUI`](/docs/threading/runOnUI) to manually schedule worklet execution on the UI thread:
```

```javascript
function myWorklet() {
  'worklet';
  console.log('Hello from the UI thread');
}

function onPress() {
  runOnUI(myWorklet)();
}
```

> You can pass data as arguments to worklets.

Reviewer 2: ```suggestion
You can pass arguments to worklets.
```

```javascript
function myWorklet(greeting) {
  'worklet';
  console.log(`${greeting} from the UI thread`);
}

function onPress() {
  runOnUI(myWorklet)('Howdy');
}
```

> Worklets are [closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures). They can access variables declared outside of it's own scope. Only variables referenced in the worklet body will be captured inside the worklet scope.

Reviewer: ```suggestion
Worklets are [closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures). They can access variables declared outside of their own scope. Only variables referenced in the worklet body will be captured inside the worklet scope.
```
Let's also mention that built-in JS functions and objects like `Intl` or `parseInt` are never captured.

Reviewer 3: I'd make it `out of their own scopes`, we are in plural 😄

```javascript
const width = 135.5;

function otherWorklet() {
  'worklet';
  console.log('Captured width is', width);
}
```

> Capturing big JavaScript objects inside of a worklet can lead to performance issues.

Reviewer: ```suggestion
Capturing large JavaScript objects inside of a worklet can lead to performance issues.
```

```javascript
const theme = {...}; // theme is a big object

function myWorklet() {
  'worklet';
  console.log(theme.color); // 🚨 referenced `color` but captured the whole `theme` object
}
```

> You can get around this problem by first assigning the prop you want to use to a separate variable .

Reviewer: ```suggestion
You can get around this problem by first assigning the prop you want to use to a separate variable.
```

```javascript
const theme = {...};
const color = theme.color;

function myWorklet() {
  'worklet';
  console.log(color); // ✅ captured only `color`
}
```

Worklets can return data within the same thread.

```javascript
function returningWorklet() {
  'worklet';
  return "I'm back"; // on the UI thread
}

function someWorklet() {
  'worklet';

>   let what = returningWorklet(); // still on the UI thread

Reviewer: ```suggestion
  const what = returningWorklet(); // still on the UI thread
```
😈

  console.log('On the UI thread, other worklet says', what);
}
```

To pass data between UI and JS thread we use [shared values](/docs/fundamentals/glossary#shared-value).

```javascript
import { useSharedValue } from 'react-native-reanimated';

function App() {
  const width = useSharedValue(100);

  function myWorklet() {
    'worklet';
    width.value += 50;
  }

>   width.value; // available on both JS and UI thread

Reviewer: Actually, you should never read or write shared value during a render.

Reviewer 3: +1, let's put it in a `React.useEffect` callback

Reviewer 2: +2

}
```

> You can run functions on the JS thread from UI thread with [`runOnJS`](/docs/threading/runOnJS). Most frequently used to call function that aren't marked with a `"worklet";` directive (i.e. most 3rd party libraries) or to update the React state.

Reviewer: ```suggestion
You can run functions on the JS thread from UI thread with [`runOnJS`](/docs/threading/runOnJS). Most frequently used to call function that aren't marked with a `"worklet";` directive (i.e. most third-party libraries) or to update the React state.
```

> You can run functions on the JS thread from UI thread with [`runOnJS`](/docs/threading/runOnJS). Most frequently used to call function that aren't marked with a `"worklet";` directive (i.e. most 3rd party libraries) or to update the React state.

Reviewer 3: Added a bit more to @tomekzaw improvement:
```suggestion
You can run functions on the JS thread from the UI thread with [`runOnJS`](/docs/threading/runOnJS). Most frequently used to call functions that aren't marked with a `'worklet';` directive (i.e. most third-party libraries) or to update the React state.
```

```javascript
import { router } from 'expo-router';
import { Gesture } from 'react-native-gesture-handler';

function App() {
  const tap = Gesture.Tap().onEnd(() => {
    // i'm a worklet too!
    // highlight-next-line
    runOnJS(router.back)();
  });
}

> ```

Reviewer 2: I'd expand this to annotate, that the function put in `runOnJS` has to be defined on the JS thread, i.e.:

```js
function worklet(){
  'worklet';
  runOnJS(() => console.log("Back to JS"))();
```

Will result in an error since the lambda function was created on the UI thread.

Author: Added in 3d50d09
